File: GO/linked-list.go
// GO LINKED LIST IMPLEMENTATION
package main
import ("fmt"
"errors")

type Node struct {
	data int
	nextNode *Node
}

type LinkedList struct {
	head *Node
}

// 1. Add a Node
func (l *LinkedList) add(inputData int) {
	newNode := Node{data: inputData, nextNode: l.head}
	l.head = &newNode
}

// 2. Search
func (l *LinkedList) search(input int) bool {
	current := l.head
	for current != nil {
		if current.data == input {
			return true
		}
		current = current.nextNode
	}
	return false
}

// 3. Find Max - USES ERROR FOR NIL
func (l *LinkedList) findMax() (int, error) {
	if l.head == nil {
		return -1, errors.New("empty list")
	}

	max := l.head.data
	current := l.head.nextNode

	for current != nil {
		if current.data > max {
			max = current.data
		}
		current = current.nextNode
	}
	return max, nil
}

// 4. Find Min
func (l *LinkedList) findMin() int {
	min := l.head.data
	current := l.head.nextNode
	for current != nil {
		if current.data < min {
			min = current.data
		}
		current = current.nextNode
	}
	return min
}

// 5. Length
func (l *LinkedList) length() int {
	count := 0
	current := l.head

	for current != nil {
		count++
		current = current.nextNode
	}
	return count
}

// 6. Find Nth from Beginning
func (l *LinkedList) findN(value int) int {
	count := 0
	current := l.head
	for current != nil {
		if count == (value - 1) {
			return current.data
		}
		count++
		current = current.nextNode
	}
	return current.data
}

// 7. Insert Ascending, assumes LL is sorted - Thank you Karinna!
func (l *LinkedList) ascInsert(inputData int) {
	newNode := Node{data: inputData}

	// if list is empty, just insert node as head
	if l.head == nil {
		l.head = &newNode
	} else if l.head.data > inputData {
		// if the head is bigger than the input data, insert node as head's nextNode
		newNode.nextNode = l.head
		l.head = &newNode
	} else {
		current := l.head

		//if list has one, smaller node: point to new node last
		if current.nextNode == nil {
			current.nextNode = &newNode
		} else {
			// otherwise, iterate through list: stop when current's next node is bigger than the input data, or it's the end of the list.
			for current.nextNode != nil && current.nextNode.data < inputData {
				current = current.nextNode
			}
			// assign newNode's nextNode if current.next's data > inputData or nil as the; reassign current's next node as newNode
			newNode.nextNode = current.nextNode
			current.nextNode = &newNode

		}
	}
}

// 8. Visit, print each node
func (l *LinkedList) visit() {
	current := l.head
	for current != nil {
		fmt.Println(current.data)
		current = current.nextNode
	}
}

// 9. Delete first found of specified value
func (l *LinkedList) delete(inputData int) {
	current := l.head
	if current.data == inputData { // head is value
		l.head = current.nextNode
		return
	}
	for current.nextNode != nil && current.nextNode.data != inputData {
		current = current.nextNode
	}
	if current.nextNode != nil {
		current.nextNode = current.nextNode.nextNode
	}
}

// 10. Reverse list
func (l *LinkedList) reverseList() {
	//declare variables
	var previous *Node = nil
	current := l.head
	var next *Node = nil

	//reverse list
	for current != nil {
		next = current.nextNode
		current.nextNode = previous
		previous = current
		current = next
	}

	//reset head
	l.head = previous
}

func main() {
	numList := &LinkedList{nil }

	//Add more nodes
	numList.add(1)
	numList.add(2)
	numList.add(3)
	numList.add(4)
	numList.visit()
	fmt.Println("-------------")

	//reverse and print list
	//numList.reverseList()
	//numList.visit()
	//fmt.Println("-------------")

	//insert ascending
	numList.reverseList()
	numList.visit()
	fmt.Println("-------------")
	numList.ascInsert(7)
	numList.ascInsert(5)
	numList.visit()
	fmt.Println("-------------")

	//delete and print list
	//numList.delete(3)
	//numList.delete(4)
	//numList.delete(1)
	//numList.visit()
	//numList.delete(2)
}

File: Ruby/linked-list.rb
# RUBY LINKED LIST IMPLEMENTATION
class Node
  attr_accessor :next
  attr_reader :value

  def initialize(input)
    @value = input
    @next = nil
  end
end

class LinkedList
  def initialize
    @head = nil
  end

  # 1. Add New Node
  def add_node(input)
    if !@head
      @head = Node.new(input)
    else
      temp = @head
      @head = Node.new(input)
      @head.next = temp
    end
  end

  # 2. Contains Value? Boolean
  def has_value?(input)
    current = @head
    while current
      return true if current.value == input
      current = current.next
    end
    return false
  end

  # 3. Max Value
  def get_max
    if !@head
      return "no content!"
    else
      max = @head.value
      current = @head
      while current
        if max < current.value
          max = current.value
        end
        current = current.next
      end
      return max
    end
  end

  # 4. Min Value
  def get_min
    if !@head
      return "no content!"
    else
      min = @head.value
      current = @head
      while current
        if min > current.value
          min = current.value
        end
        current = current.next
      end
      return min
    end
  end

  # 5. Count Nodes
  def get_size
    count = 0
    current = @head
    while current
      count += 1
      current = current.next
    end
    return count
  end

  # 6. Nth Node from the Beginning; Index Starts at 0
  def nth_node(input)
    current = @head
    (input).times do
      return "not enough content!" if !current
      current = current.next
    end
    return current.value
  end

  # 7. Insert Node Asc Order
  def asc_insert(input)
    if !@head
      #list is empty
      @head = Node.new(input)
    elsif @head.value > input
      #input smaller than head value
      add_node(input)
    else
      #input larger than head value
      current = @head
      if !current.next
        #if list has 1 item
        current.next = Node.new(input)
      else
        #if list has more than 1 item iterate until current's next node is bigger than input
        while current.next && current.next.value < input
          current = current.next
        end
        ourNode = Node.new(input)
        ourNode.next = current.next
        current.next = ourNode
      end
    end
  end

  # 8. Print Values Separated by a Space
  def visit
    return puts "no content!" if !@head

    current = @head
    while current
      print current.value.to_s + " "
      current = current.next
    end
    puts ""
  end

  # 9. Delete First Found of Value
  def delete(input)
    #TODO what if the input cannot be found?
    #if empty
    return puts "no content!" if !@head
    current = @head
    #if 1st node has value
    if current.value == input
      if current.next
        #if there is 2nd node in list
        @head = current.next
        return
      else
        #if only 1 node
        @head = nil
        return
      end
    end
      #keep looking for the value
    while current
      if current.next.value == input && current.next.next
        #if value is found
        current.next = current.next.next
        return
      end
      current = current.next
    end
  end

  # 10. Reverse List
  def reverse
    #no content or only 1 node
    return puts "insufficient content!" if !@head || !@head.next

    current = @head
    previous = nil
    while current
      nextNode = current.next
      current.next = previous
      previous = current
      current = nextNode
    end
    @head = previous
  end

end

# --------------------------------------------------
# INITIALIZE
# myList = LinkedList.new()
# myList.visit
# puts "--------------"

# CHECK METHODS
# 1 Add
# myList.add_node(7)
# myList.add_node(6)
# myList.add_node(1)

#2 Contains?
# puts myList.has_value?(1)
# puts myList.has_value?(5)

#3 Max, #4 Min
# puts myList.get_max
# puts myList.get_min

#5 Size
# puts myList.get_size

#6 Nth Node
# myList.visit
# puts myList.nth_node(0)
# puts myList.nth_node(1)
# puts myList.nth_node(5)

#7 Insert Asc
# puts myList.get_size
# myList.asc_insert(6)
# puts myList.get_size

#8. Visit
# myList.visit

#9 Delete
# myList.visit
# myList.delete(6)
# myList.visit
# puts myList.get_size
# myList.delete(1)
# myList.visit
# puts myList.get_size
# myList.delete(7)
# myList.visit
# puts myList.get_size

#10 Reverse
# myList.visit
# myList.reverse
# myList.visit
# puts "--------------"
# myList.delete(6)
# myList.delete(7)
# myList.reverse
# myList.visit
